lib/vim-state.coffee

```diff
@@ -654,6 +662,16 @@ class VimState
     text = @getRegister(name)?.text
     @editor.insertText(text) if text?
 
+  ensureCursorIsWithinLine: (cursor) =>
+    return if @processing or @mode is 'visual' or @mode is 'insert'
+
+    if cursor.cursor? then cursor = cursor.cursor # when called by @editor.onDidChangeCursorPosition
```

Comment:
I think this would be a little more straightforward if you removed this line, and when observing the cursor position did this:

```coffee
@editor.onDidChangeCursorPosition ({cursor}) => @ensureCursorIsWithinLine(cursor)
```